bpo-33409: Clarify PEP 538/540 relationship

ncoghlan opened the pull request:
While locale coercion and UTF-8 mode turned out to
be complementary ideas rather than competing ones,
it isn't immediately obvious why it's useful to
have both, or how they interact at runtime.

This updates both the Python 3.7 What's New doc
and the PYTHONCOERCECLOCALE and PYTHONUTF8
documentation in an attempt to clarify that
relationship.


<!-- issue-number: bpo-33409 -->
https://bugs.python.org/issue33409
<!-- /issue-number -->

ncoghlan committed:
bpo-33409: Clarify PEP 538/540 relationship

While locale coercion and UTF-8 mode turned out to
be complementary ideas rather than competing ones,
it isn't immediately obvious why it's useful to
have both, or how they interact at runtime.

This updates both the Python 3.7 What's New doc
and the PYTHONCOERCECLOCALE and PYTHONUTF8
documentation in an attempt to clarify that
relationship.

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -819,6 +819,16 @@ conflict.
    activates, or else if a locale that *would* have triggered coercion is
    still active when the Python runtime is initialized.
 
+   Note that setting ``LC_ALL`` also implicitly turns of locale coercion, as
```
Turn off? Two F?

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -834,10 +844,23 @@ conflict.
-   If set to ``1``, enable the UTF-8 mode. If set to ``0``, disable the UTF-8
-   mode. Any other non-empty string cause an error.
+   If set to ``1``, enable the UTF-8 mode, such that the interpreter uses
+   ``UTF-8`` as the text encoding for system interfaces, regardless of the
```
It also sets stdin and stdout error handler to surrogateescape.

vstinner commented on Doc/whatsnew/3.7.rst:
```diff
@@ -231,8 +241,17 @@ The forced UTF-8 mode can be used to change the text handling behavior in
-:ref:`whatsnew37-pep538` for details.
+While :pep:`540`'s UTF-8 mode has the benefit of working regardless of which
+locales are available on the running system, it has the downside of having no
+effect on extension modules, child processes running non-Python applications,
```
If you want to enumerate things not affected by the utf8 mode, I would also suggest to mention C libraries such as readline.

vstinner reviewed (commented):
A few comments. Thanks for your PR :-)

ncoghlan committed:
Further doc updates

- mention GNU readline in the PEP 540 What's New section
- mention UTF-8 mode change the stdin and stderr error handler
- improve wording conistenct between the PYTHONCOERCECLOCALE
  and PYTHONUTF8MODE docs when they cover the same thing (mostly
  related to legacy locale detection and setting the standard
  stream error handler)
- move the reference to LC_ALL turning off locale coercion into
  the description of how locale coercion activates in the first
  place
- port the full description of the UTF-8 mode behaviour changes
  from PEP 540 into the PYTHONUTF8 documentation

ncoghlan committed:
Fix markup

ncoghlan committed:
Further cleanups

- more explicit -X utf8 description
- GNU readline reads the current locale, not the environment
- other minor tweaks

tonybaloney commented on Doc/using/cmdline.rst:
```diff
@@ -438,8 +438,10 @@ Miscellaneous options
-   * ``-X utf8`` enables the UTF-8 mode, whereas ``-X utf8=0`` disables the
-     UTF-8 mode. See :envvar:`PYTHONUTF8` for more details.
+   * ``-X utf8`` enables UTF-8 mode for operating system interfaces, overriding
+     the default locale-aware mode. ``-X utf8=0`` explicitly disables UTF-8
```
Much clearer 👍

ncoghlan committed:
Markup fix
